scipy/stats/_hypotests.py

```diff
@@ -663,3 +666,361 @@ def somersd(x, y=None):
         raise ValueError("x must be either a 1D or 2D array")
     d, p = _somers_d(table)
     return SomersDResult(d, p, table)
+
+
+def _compute_log_combinations(n):
+    """Compute all log combination of C(n, k)"""
+    gammaln_arr = gammaln(np.arange(n + 1) + 1)
+    return gammaln(n + 1) - gammaln_arr - gammaln_arr[::-1]
+
+
+BarnardExactResult = make_dataclass(
+    "BarnardExactResult", [("statistic", float), ("pvalue", float)]
+)
+
+
+def barnard_exact(table, alternative="two-sided", pooled=True, n=32):
+    r"""Perform a Barnard exact test on a 2x2 contingency table.
+
+    Parameters
+    ----------
+    table : array_like of ints
+        A 2x2 contingency table.  Elements should be non-negative integers.
+    alternative : {'two-sided', 'less', 'greater'}, optional
+        Defines the alternative hypothesis.
+        The following options are available (default is 'two-sided'):
+
+        * 'two-sided'
+        * 'less': one-sided
+        * 'greater': one-sided
+
+    pooled : bool, optional
+        Whether to compute score statistic with pooled variance (Student's
+        t-test) or unpooled variance (Welch's t-test). Default is ``True`` :
+        Statistic test is computed using pooled variance.
+
+    n : int, optional
+        Number of sampling points used in the construction of the sampling
+        method. Note that this argument must be a power of two since we are
+        using `scipy.stats.qmc.Sobol`.
```

Comment:
I was thinking on putting a low number of `n` (like 3, 4) and then using it as a power of two, to avoid the user to get wrong. What do you think ? 
Basically, this would me to pass to `shgo` function, the parameter `n = 2**n`, the latter designing our n. 
line 924 : 
```Python
    result = shgo(
        _get_binomial_log_p_value_with_nuisance_param,
        args=(x1_sum_x2, x1_sum_x2_log_comb, index_arr),
        bounds=((0, 1),),
        n=2**n,  # Need to be a power of two since it is used by sobol
        sampling_method="sobol",
    )
```

Reply:
In fact, the sobol seq is generated by `Sobol.i4_sobol_generate` in `scipy.optimize._shgo_lib.sobol_seq`, with any number of point. If the number of point is a power of two minus one (15, 31, 63, 127…) the sequence is equidistributed with `d=1`, but the number of point should not be forced to be a power of two minus one (or a power of two).

See:

```
In [1]: from scipy.optimize._shgo_lib.sobol_seq import Sobol
   ...: S = Sobol()

In [2]: import numpy as np

In [3]: S.i4_sobol_generate(1, 12)
Out[3]:
array([[0.5   ],
       [0.75  ],
       [0.25  ],
       [0.375 ],
       [0.875 ],
       [0.625 ],
       [0.125 ],
       [0.1875],
       [0.6875],
       [0.9375],
       [0.4375],
       [0.3125]])

In [4]: S.i4_sobol_generate(1, 18)
Out[4]:
array([[0.5    ],
       [0.75   ],
       [0.25   ],
       [0.375  ],
       [0.875  ],
       [0.625  ],
       [0.125  ],
       [0.1875 ],
       [0.6875 ],
       [0.9375 ],
       [0.4375 ],
       [0.3125 ],
       [0.8125 ],
       [0.5625 ],
       [0.0625 ],
       [0.09375],
       [0.59375],
       [0.84375]])

In [5]: S.i4_sobol_generate(1, 21)
Out[5]:
array([[0.5    ],
       [0.75   ],
       [0.25   ],
       [0.375  ],
       [0.875  ],
       [0.625  ],
       [0.125  ],
       [0.1875 ],
       [0.6875 ],
       [0.9375 ],
       [0.4375 ],
       [0.3125 ],
       [0.8125 ],
       [0.5625 ],
       [0.0625 ],
       [0.09375],
       [0.59375],
       [0.84375],
       [0.34375],
       [0.46875],
       [0.96875]])

In [6]: minmax = lambda x: (x.min(), x.max())

In [10]: minmax(np.diff(np.sort(S.i4_sobol_generate(1, 15).flatten())))
Out[10]: (0.0625, 0.0625)

In [11]: minmax(np.diff(np.sort(S.i4_sobol_generate(1, 16).flatten())))
Out[11]: (0.03125, 0.0625)

In [12]: minmax(np.diff(np.sort(S.i4_sobol_generate(1, 31).flatten())))
Out[12]: (0.03125, 0.03125)

In [13]: minmax(np.diff(np.sort(S.i4_sobol_generate(1, 32).flatten())))
Out[13]: (0.015625, 0.03125)

In [24]: minmax(np.diff(np.sort(S.i4_sobol_generate(1, 63).flatten())))
Out[24]: (0.015625, 0.015625)

In [25]: minmax(np.diff(np.sort(S.i4_sobol_generate(1, 64).flatten())))
Out[25]: (0.0078125, 0.015625)
```

So:
 - I propose to set the default `n` to a power of two minus one.
 - Keep `n` at it is now, without enforcing `n` to be a power of two or a power of two minus one.

Reply:
Whatever happens is fine. We should just make sure that the documentation and behavior match. 

> I would not do anything here

That would be fine with me if gh-13469 merges first. If this will merge first, I'd suggest adding the behavior and a test here and removing it in gh-13469.

Reply:
Ok, so I am not exactly sure of what I am suppose to do here. I believe this depends on how long before you think this PR or gh-13469 will be merged.

Reply:
@mdhaber in #13469, the rounding to a power of 2 is taken care of. So there would be nothing to validate here as any non power would be converted later by shgo. So any validation and tests for this here would break or be rendered useless as soon as the new version would be merged.

#13469 is ready on my side and just waits for more reviews or integration.

Reply:
> @mdhaber in #13469, the rounding to a power of 2 is taken care of

Then do you agree that rather than the current

> Note that this argument **must** be a power of two since we are using `scipy.stats.qmc.Sobol`.

something like

> Note that this argument will automatically be converted to the next higher power of 2 since `scipy.stats.qmc.Sobol` is used to select sample points.

would be more accurate?

Reply:
> would more accurately reflect the behavior of this function?

I realize now that it does not accurately reflect the behavior of this function yet, as gh-13469 has not merged yet.

> Ok, so I am not exactly sure of what I am suppose to do here.

For now, I would remove "Note that this argument must be a power of two since we are using scipy.stats.qmc.Sobol." As I understand it, `shgo` does not currently use `scipy.stats.qmc.Sobol`, nor does it require that `n` be a power of 2, so it seems that this statement can just be removed.

When gh-13469 merges, this can change to something like "Note that this argument will automatically be converted to the next higher power of 2 since scipy.stats.qmc.Sobol is used to select sample point." But - again, if I understand correctly - that is not accurate now, so unless you want this PR to depend on the merge of gh-13469, removing the statement seems to me the best course.